pgx/src/datum/range.rs

```diff
@@ -0,0 +1,332 @@
+    pub fn from_range_values(
+        lower_val: Option<T>,
+        upper_val: Option<T>,
+        lower_inc: bool,
+        upper_inc: bool,
+    ) -> Self {
+        let mut lower_bound = RangeBound {
+            lower: true,
+            inclusive: lower_inc,
+            ..Default::default()
+        };
+
+        let mut upper_bound = RangeBound {
+            lower: false,
+            inclusive: upper_inc,
+            ..Default::default()
+        };
+
+        match lower_val {
+            Some(lower_val) => {
+                lower_bound.val = lower_val
+                    .into_datum()
+                    .expect("Couldn't convert lower_val to Datum");
+            }
+            None => {
+                lower_bound.infinite = true;
+            }
+        }
+
+        match upper_val {
+            Some(upper_val) => {
+                upper_bound.val = upper_val
+                    .into_datum()
+                    .expect("Couldn't convert upper_val to Datum");
+            }
+            None => {
+                upper_bound.infinite = true;
+            }
+        }
+
+        RangeData::from_range_bounds(lower_bound, upper_bound)
+    }
+}
+
+impl<T> From<Range<T>> for RangeData<T>
+where
+    T: FromDatum + IntoDatum + RangeSubType,
+{
+    /// ## Safety
+    /// Requires that:
+    /// - range.range_type is valid pointer to RangeType PG struct
+    fn from(range: Range<T>) -> Self {
+        let mut lower_bound: RangeBound = Default::default();
+        let mut upper_bound: RangeBound = Default::default();
+        let mut is_empty = false;
+
+        unsafe {
+            let typecache = pg_sys::lookup_type_cache(
+                (*(range.range_type)).rangetypid,
+                pg_sys::TYPECACHE_RANGE_INFO as i32,
+            );
+
+            pg_sys::range_deserialize(
+                typecache,
+                range.range_type,
+                &mut lower_bound,
+                &mut upper_bound,
+                &mut is_empty,
+            );
+        }
+
+        RangeData::from_range_bounds_internal(lower_bound, upper_bound, is_empty)
+    }
+}
+
+impl<T> From<RangeData<T>> for Range<T>
+where
+    T: FromDatum + IntoDatum + RangeSubType,
+{
+    fn from(range_data: RangeData<T>) -> Self {
+        unsafe {
+            let typecache =
+                pg_sys::lookup_type_cache(T::range_type_oid(), pg_sys::TYPECACHE_RANGE_INFO as i32);
+
+            let mut lower_bound = range_data.lower;
+            let mut upper_bound = range_data.upper;
+
+            let range_type = pg_sys::make_range(
+                typecache,
+                &mut lower_bound,
+                &mut upper_bound,
+                range_data.is_empty,
+            );
+
+            Range::<T>::from_pg(range_type as *mut pg_sys::varlena, range_type)
+        }
+    }
+}
+
+/// This trait allows a struct to be a valid subtype for a RangeType
+pub trait RangeSubType {
```

Comment:
Using "subtype" is potentially confounding. In Rust, this trait can be used in a generic bound. Rust does not have "subtypes" as usually considered, except in the form of the lifetime system (lifetime subtyping allowing different lifetimes to satisfy the same constraint, as long as each "`'a` outlives `'b`" and so on constraint is met).

If you mean that it satisfies a Postgres subtype relationship, then that is another thing, in which case this trait should probably be sealed.

If you mean that they all implement the same behavior, then they should just implement the same behavior via the trait.

Reply:
If it is possible for the trait implementation to cause unsound results (e.g. if a wrong answer from `range_type_oid()` could, when used with other functions, cause unsoundness, or if implementing this for the _wrong_ type can cause unsoundness) then the bound cannot be fully "trusted" unless this is marked as a "trusted" trait (i.e. is an `unsafe Trait`). Note this does not mean the trait's functions must be `unsafe`: any `fn` of an `unsafe Trait` may be itself safe or `unsafe` to call, the `unsafe` marker on the trait instead says the entire trait has a compiler-unverified invariant and thus must be implemented correctly by the programmer.

Reply:
I've made RangeSubType an unsafe trait